File: lib/49.group_anagrams.dart
void main() {
  print(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]));
  print(groupAnagrams([""]));
  print(groupAnagrams(["a"]));
}

List<List<String>> groupAnagrams(List<String> strs) {
  Map<String, List<String>> map = {};

  strs.forEach((str) {
    final partStr = str.split('')..sort();
    final key = partStr.join();

    if (map.containsKey(key)) {
      map[key]?.add(str);
    } else {
      map[key] = [str];
    }
  });

  return map.values.toList();
}

File: src/49.group_anagrams.kt
fun main() {
    println(groupAnagrams(arrayOf("eat", "tea", "tan", "ate", "nat", "bat")))
    println(groupAnagrams(arrayOf("")))
    println(groupAnagrams(arrayOf("a")))
}

fun groupAnagrams(strs: Array<String>): List<List<String>> {
    val map = mutableMapOf<String, MutableList<String>>()

    strs.forEach { str ->
        val partStr = str.split("").sorted()
        val key = partStr.joinToString()

        if (map.containsKey(key)) {
            map[key]?.add(str)
        } else {
            map[key] = mutableListOf(str)
        }
    }

    return map.values.toList()
}